File: java_solutions/86.partition-list.java
/*
 * @lc app=leetcode id=86 lang=java
 *
 * [86] Partition List
 */

class Solution {
    public ListNode partition(ListNode head, int x) {
        return new Object[]{java.util.stream.Stream.iterate(head, n -> n != null, n -> n.next).toArray(ListNode[]::new)} instanceof Object[] arr 
            ? java.util.stream.Stream.concat(
                java.util.stream.Stream.of((ListNode[])arr[0]).filter(n -> n.val < x), 
                java.util.stream.Stream.of((ListNode[])arr[0]).filter(n -> n.val >= x)
              ).reduce((a, b) -> (a.next = b))
               .map(last -> (last.next = null) == null ? 
                  java.util.stream.Stream.concat(
                      java.util.stream.Stream.of((ListNode[])arr[0]).filter(n -> n.val < x), 
                      java.util.stream.Stream.of((ListNode[])arr[0]).filter(n -> n.val >= x)
                  ).findFirst().orElse(null) 
                  : null
               ).orElse(null) 
            : null;
    }
}

File: kotlin_solutions/86.partition-list.kt
/*
 * @lc app=leetcode id=86 lang=kotlin
 *
 * [86] Partition List
 */

class Solution { fun partition(head: ListNode?, x: Int): ListNode? = ListNode(0).let { smallHead -> ListNode(0).let { bigHead -> generateSequence(Triple(head, smallHead, bigHead)) { (curr, currSmall, currBig) -> curr?.let { if (it.`val` < x) Triple(curr.next, currSmall.apply { next = curr }.next, currBig) else Triple(curr.next, currSmall, currBig.apply { next = curr }.next) } }.last().let { (_, currSmall, currBig) -> currBig.apply { next = null } .let { currSmall.apply { next = bigHead.next } } .let { smallHead.next } } } } }
